MerlinAU.sh

```diff
@@ -291,38 +291,79 @@ _WaitForYESorNO_()
+## Modified by Martinski W. [2024-Nov-14] ##
+##----------------------------------------##
+readonly LockFilePath="/tmp/var/${ScriptFNameTag}.LOCK"
+LockFileMaxAgeSecs=600  #10-minutes#
+LockWaitTimeoutSecs=120
+
 _ReleaseLock_() { rm -f "$LockFilePath" ; }
 
+if [ $# -eq 0 ] || [ -z "$1" ]
+then  #Interactive Mode#
+    LockFileMaxAgeSecs=1200
+    LockWaitTimeoutSecs=10
+elif [ "$1" = "addCronJob" ]
+then  #Special Case#
+    LockFileMaxAgeSecs=1200
+    LockWaitTimeoutSecs=600
+fi
+
+##----------------------------------------##
+## Modified by Martinski W. [2024-Nov-15] ##
+##----------------------------------------##
 _AcquireLock_()
 {
    if [ ! -f "$LockFilePath" ]
    then
        echo "$$" > "$LockFilePath"
        return 0
    fi
+   local lockFileSecs  ageOfLockSecs  oldPID  retCode  waitTimeoutSecs
 
-   ageOfLockSecs="$(($(date +%s) - $(date +%s -r "$LockFilePath")))"
-   if [ "$ageOfLockSecs" -gt "$LockFileMaxSecs" ]
-   then
-       Say "Stale lock found (older than $LockFileMaxSecs secs.) Reset lock file."
-       oldPID="$(cat "$LockFilePath")"
-       if [ -n "$oldPID" ] && kill -EXIT "$oldPID" 2>/dev/null && \
-          pidof "$ScriptFileName" | grep -qow "$oldPID"
-       then
-           kill -TERM "$oldPID" ; wait "$oldPID"
-       fi
-       rm -f "$LockFilePath"
-       echo "$$" > "$LockFilePath"
-       return 0
-   else
-       Say "${REDct}**ERROR**${NOct}: The shell script '${ScriptFileName}' is already running [Lock file: $ageOfLockSecs secs.]"
-       return 1
-   fi
+   retCode=1
+   waitTimeoutSecs=0
+
+   while true
+   do
+      if [ -f "$LockFilePath" ]
+      then
+          oldPID="$(cat "$LockFilePath")"
+          lockFileSecs="$(date +%s -r "$LockFilePath")"
+      else
+          echo "$$" > "$LockFilePath"
+          retCode=0
+          break
+      fi
+
+      ageOfLockSecs="$(($(date +%s) - lockFileSecs))"
+      if [ "$ageOfLockSecs" -gt "$LockFileMaxAgeSecs" ]
+      then
+          Say "Stale Lock Found (older than $LockFileMaxAgeSecs secs.) Resetting lock file..."
+          if [ -n "$oldPID" ] && kill -EXIT "$oldPID" 2>/dev/null && \
+              pidof "$ScriptFileName" | grep -qow "$oldPID"
+          then
+              kill -TERM "$oldPID" ; wait "$oldPID"
+          fi
+          echo "$$" > "$LockFilePath"
+          retCode=0
+          break
+      elif [ "$waitTimeoutSecs" -le "$LockWaitTimeoutSecs" ]
+      then
+          if [ "$waitTimeoutSecs" -eq 0 ] || [ "$((waitTimeoutSecs % 10))" -eq 0 ]
+		  then
+              Say "Lock Found [Age: $ageOfLockSecs secs.] Waiting for script [PID=$oldPID] to exit [Timer: $waitTimeoutSecs secs.]"
+          fi
+          sleep 2
+          waitTimeoutSecs="$((waitTimeoutSecs + 2))"
+      else
+          Say "${REDct}**ERROR**${NOct}: The shell script ${ScriptFileName} [PID=$oldPID] is already running [Lock Age: $ageOfLockSecs secs.]"
+          retCode=1
+          break
+      fi
+   done
+   return "$retCode"
```

Comment:
I like the fact that with this; if the lock is active and not stale, the script does not exit immediately. instead it enters a waiting phase where it periodically checks if the lock has been released. :)